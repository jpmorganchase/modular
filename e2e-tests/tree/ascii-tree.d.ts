declare module 'ascii-tree' {
  export function generate(input: string): string;
}
