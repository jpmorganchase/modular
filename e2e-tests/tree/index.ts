import fs from 'fs';
import path from 'path';
import asciiTree from 'ascii-tree';
import hash from '@emotion/hash';

function times(str: string, length: number) {
  return Array.from({ length }, () => str).join('');
}

const defaultIgnores = ['node_modules', '.git', '.DS_Store'];
const defaultHashIgnores = [
  // adding lockfiles here because it can be different
  // on different runs; since we install the latest versions
  // of some packages when making a repository
  'yarn.lock',
  'package-lock.json',
  // adding package.json files since versions within these get bumped
  'package.json',
];

function tree(
  _path: string,
  level = 1,
  options = {
    ignores: defaultIgnores,
    hashIgnores: defaultHashIgnores,
  },
): string {
  const stat = fs.statSync(_path);
  if (stat.isDirectory()) {
    const children = fs.readdirSync(_path);
    const dirArr = _path.split('/');
    const dir = dirArr[dirArr.length - 1];
    // todo - should these be sorted?
    // todo - handle symlinks, etc
    return `${times('#', level)}${dir}\n${children
      .filter((child: string) => !options.ignores.includes(child))
      .map((child: string) => tree(path.join(_path, child), level + 1, options))
      .join('\n')}`;
  } else {
    return [
      `${times('#', level)}${path.basename(_path)}`,
      options.hashIgnores.includes(path.basename(_path))
        ? undefined
        : `#${hash(fs.readFileSync(_path, 'utf8'))}`,
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export default function generate(_path: string): string {
  return asciiTree.generate(tree(_path));
}
