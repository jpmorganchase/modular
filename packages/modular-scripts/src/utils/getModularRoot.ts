import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
