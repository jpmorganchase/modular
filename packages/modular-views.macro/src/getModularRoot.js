// keep this in sync with modular-scripts/src/getModularRoot
const path = require('path');
const fs = require('fs-extra');
const findUp = require('find-up');

function isModularRoot(packageJson) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

module.exports = function getModularRoot() {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
};
