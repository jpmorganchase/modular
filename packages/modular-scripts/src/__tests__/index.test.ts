import execa from 'execa';
import rimraf from 'rimraf';
import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';

import {
  getDocument,
  getQueriesForElement,
  queries,
} from 'pptr-testing-library';

import getModularRoot from '../utils/getModularRoot';

const modularRoot = getModularRoot();

const START_APP_TIMEOUT = 60 * 1000;

interface DevServer {
  kill: () => void;
}
// eslint-disable-next-line @typescript-eslint/unbound-method
const { getNodeText } = queries;

// These tests must be executed sequentially with `--runInBand`.
jest.setTimeout(10 * 60 * 1000);

const packagesPath = path.join(getModularRoot(), 'packages');

function modular(str: string, opts: Record<string, unknown> = {}) {
  return execa('yarnpkg', ['modular', ...str.split(' ')], {
    cwd: modularRoot,
    cleanup: true,
    ...opts,
  });
}

async function startApp(appPath: string): Promise<DevServer> {
  const devServer = modular(`start ${appPath}`, {
    cleanup: true,
  });

  await new Promise((resolve, reject) => {
    if (!devServer.stdout) {
      return reject(
        new Error(
          'The dev server could not produce any output on /dev/stdout.',
        ),
      );
    }

    if (!devServer.stderr) {
      return reject(
        new Error(
          'The dev server could not produce any output on /dev/stderr.',
        ),
      );
    }

    // If the Promise has completed we need to bail out of any further handlers that
    // are executed, because otherwise we can end up trying to log after tests are done.
    // We use a 'completed' variable to do this -- in an ideal world we'd cleanup
    // the listeners.
    //
    // See: https://github.com/jpmorganchase/modular/pull/107#discussion_r493791918
    let completed = false;

    const startAppTimeout = setTimeout(() => {
      if (completed) return;

      completed = true;
      reject(
        new Error(
          `The app at ${appPath} never started within the configured ${START_APP_TIMEOUT}ms timeout period.`,
        ),
      );
    }, START_APP_TIMEOUT);

    devServer.stdout.on('data', (data: Buffer) => {
      if (completed) return;

      const output = data.toString();
      if (/Something is already running on port (\d+)./.test(output)) {
        clearTimeout(startAppTimeout);

        completed = true;
        return reject(new Error(output));
      }
      if (/Compiled successfully!/.test(output)) {
        clearTimeout(startAppTimeout);

        completed = true;
        return resolve(true);
      }
    });

    devServer.stderr.on('data', (data: Buffer) => {
      if (completed) return;

      const output = data.toString();

      console.error(output);

      clearTimeout(startAppTimeout);

      completed = true;
      return reject(new Error(output));
    });

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    devServer.on('error', (err: Error) => {
      if (completed) return;

      console.error(err);

      clearTimeout(startAppTimeout);

      completed = true;
      reject(err);
    });
  });

  // We can't return the raw execa child process,
  // because we need to `await` until the dev server
  // starts above, but don't want to wait until the
  // process itself finishes.
  return {
    kill: () => {
      devServer.kill('SIGKILL');
      return devServer;
    },
  };
}

function cleanup() {
  rimraf.sync(path.join(packagesPath, 'sample-app'));
  rimraf.sync(path.join(packagesPath, 'sample-view'));
  rimraf.sync(path.join(packagesPath, 'sample-package'));
  rimraf.sync(path.join(packagesPath, 'nested'));
  rimraf.sync(path.join(modularRoot, 'dist'));
  // run yarn so yarn.lock gets reset
  return execa.sync('yarnpkg', ['--silent'], {
    cwd: modularRoot,
  });
}

beforeAll(cleanup);
afterAll(cleanup);

describe('modular-scripts', () => {
  it('can add an app', async () => {
    await modular(
      'add sample-app --unstable-type app --unstable-name sample-app',
      { stdio: 'inherit' },
    );

    // Let's replace the App module with something of our own
    // with a test specific element we can introspect
    await fs.copyFile(
      path.join(__dirname, 'TestApp.test-tsx'),
      path.join(packagesPath, 'sample-app', 'src', 'App.tsx'),
    );

    expect(tree(path.join(packagesPath, 'sample-app'))).toMatchInlineSnapshot(`
      "sample-app
      ├─ package.json
      ├─ public
      │  ├─ favicon.ico #6pu3rg
      │  ├─ index.html #1wohq3p
      │  ├─ logo192.png #1nez7vk
      │  ├─ logo512.png #1hwqvcc
      │  ├─ manifest.json #19gah8o
      │  └─ robots.txt #1sjb8b3
      ├─ src
      │  ├─ App.css #1o0zosm
      │  ├─ App.tsx #igvgtx
      │  ├─ __tests__
      │  │  └─ App.test.tsx #16urcos
      │  ├─ index.css #o7sk21
      │  ├─ index.tsx #zdn6mw
      │  ├─ logo.svg #1okqmlj
      │  └─ react-app-env.d.ts #1dm2mq6
      └─ tsconfig.json #6rw46b"
    `);
  });

  it('can start an app', async () => {
    // Ok, so. Sunil's decided to get the new M1 MacBook Air. Some software doesn't run on it
    // well yet. Particularly the puppeteer npm package failes to install and run
    // (see https://github.com/puppeteer/puppeteer/issues/, issues #6634 and #6641,
    // possible fix in pull #6495)

    // Because of this, he's marked puppeteer in optionalDependencies, so it's failure to install
    // doesn't block everything else. Further, because this particular test is already flaky,
    // it's disabled when running locally. However, because it fails to install, it causes
    // typescript and eslint failures. Hence the need to disable those errors for now.

    // It's Sunil's responsibility to fix this when better, so shout at him if he doesn't.

    /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */

    // This seems to be leaving hanging processes locally,
    // so marking this test as a no-op for now. Sigh.
    if (!process.env.CI) {
      return;
    }

    const puppeteer = require('puppeteer');

    // @ts-expect-error FIXME
    let browser: puppeteer.Browser | undefined;
    let devServer: DevServer | undefined;
    try {
      await fs.copyFile(
        path.join(__dirname, 'TestApp.test-tsx'),
        path.join(packagesPath, 'sample-app', 'src', 'App.tsx'),
      );

      browser = await puppeteer.launch(
        process.env.CI
          ? {
              headless: true,
              args: ['--no-sandbox', '--disable-setuid-sandbox'],
            }
          : {},
      );
      devServer = await startApp('sample-app');

      const page = await browser.newPage();
      await page.goto('http://localhost:3000', {});

      // eslint-disable-next-line @typescript-eslint/unbound-method
      const { getByTestId, findByTestId } = getQueriesForElement(
        await getDocument(page),
      );

      await findByTestId('test-this');

      // eslint-disable-next-line testing-library/no-await-sync-query
      expect(await getNodeText(await getByTestId('test-this'))).toBe(
        'this is a modular app',
      );
    } finally {
      if (browser) {
        await browser.close();
      }
      if (devServer) {
        // this is the problematic bit, it leaves hanging node processes
        // despite closing the parent process. Only happens in tests!
        devServer.kill();
      }
    }

    /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
  });

  it('can add a view', async () => {
    await modular(
      'add sample-view --unstable-type view --unstable-name sample-view',
      { stdio: 'inherit' },
    );
    expect(tree(path.join(packagesPath, 'sample-view'))).toMatchInlineSnapshot(`
      "sample-view
      ├─ README.md #11adaka
      ├─ package.json
      └─ src
         ├─ __tests__
         │  └─ index.test.tsx #slarlz
         └─ index.tsx #fxrie0"
    `);
  });

  it('can start a view', async () => {
    /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
    if (!process.env.CI) {
      return;
    }

    const puppeteer = require('puppeteer');

    // @ts-expect-error FIXME
    let browser: puppeteer.Browser | undefined;
    let devServer: DevServer | undefined;
    try {
      const targetedView = 'sample-view';
      await fs.copyFile(
        path.join(__dirname, 'TestView.test-tsx'),
        path.join(packagesPath, targetedView, 'src', 'index.tsx'),
      );

      browser = await puppeteer.launch(
        process.env.CI
          ? {
              headless: true,
              args: ['--no-sandbox', '--disable-setuid-sandbox'],
            }
          : {},
      );
      devServer = await startApp(targetedView);

      const page = await browser.newPage();
      await page.goto('http://localhost:3000', {});

      // eslint-disable-next-line @typescript-eslint/unbound-method
      const { getByTestId, findByTestId } = getQueriesForElement(
        await getDocument(page),
      );

      await findByTestId('test-this');

      // eslint-disable-next-line testing-library/no-await-sync-query
      expect(await getNodeText(await getByTestId('test-this'))).toBe(
        'this is a modular view',
      );
    } finally {
      if (browser) {
        await browser.close();
      }
      if (devServer) {
        // this is the problematic bit, it leaves hanging node processes
        // despite closing the parent process. Only happens in tests!
        devServer.kill();
      }
    }

    /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
  });

  it('can add a package', async () => {
    await modular(
      'add sample-package --unstable-type package --unstable-name sample-package',
      {
        stdio: 'inherit',
      },
    );
    expect(tree(path.join(packagesPath, 'sample-package')))
      .toMatchInlineSnapshot(`
      "sample-package
      ├─ README.md #1jv3l2q
      ├─ package.json
      └─ src
         ├─ __tests__
         │  └─ index.test.ts #1qvvmz7
         └─ index.ts #1woe74n"
    `);
  });

  it('can add a nested package', async () => {
    await modular(
      'add nested/sample-nested-package --unstable-type package --unstable-name @nested/sample-package',
      {
        stdio: 'inherit',
      },
    );
    expect(tree(path.join(packagesPath, 'nested/sample-nested-package')))
      .toMatchInlineSnapshot(`
      "sample-nested-package
      ├─ README.md #1jv3l2q
      ├─ package.json
      └─ src
         ├─ __tests__
         │  └─ index.test.ts #1qvvmz7
         └─ index.ts #1woe74n"
    `);
  });

  it('can execute tests', async () => {
    const output = await modular(
      'test sample-app sample-package sample-view sample-nested-package --watchAll false',
      {
        all: true,
        reject: false,
        env: {
          CI: 'true',
        },
      },
    );

    // TODO: Passing CI=true *should* remove all the coloring stuff,
    // it's weird that it doesn't. To workaround it, I've manually
    // removed those tokens from the string for the snapshot test.
    // Open to suggestions/fixes.

    // eslint-disable-next-line no-control-regex
    const cleanedOutput = output.all?.replace(/|\[\d+./gm, '');

    expect(cleanedOutput).toContain(
      'PASS packages/sample-app/src/__tests__/App.test.tsx',
    );
    expect(cleanedOutput).toContain(
      'PASS packages/sample-view/src/__tests__/index.test.tsx',
    );
    expect(cleanedOutput).toContain(
      'PASS packages/sample-package/src/__tests__/index.test.ts',
    );
    expect(cleanedOutput).toContain(
      'PASS packages/nested/sample-nested-package/src/__tests__/index.test.ts',
    );
  });

  it('can build libraries', async () => {
    // cleanup anything built previously
    rimraf.sync(path.join(modularRoot, 'dist'));

    // build a view
    await modular('build sample-view', { stdio: 'inherit' });
    // build a package too, but preserve modules
    await modular('build sample-package --preserve-modules', {
      stdio: 'inherit',
    });
    // build the nested package
    await modular('build nested/sample-nested-package', {
      stdio: 'inherit',
    });

    expect(
      await fs.readJson(
        path.join(modularRoot, 'dist', 'sample-package', 'package.json'),
      ),
    ).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {},
        "files": Array [
          "/dist-cjs",
          "/dist-es",
          "/dist-types",
          "README.md",
        ],
        "license": "UNLICENSED",
        "main": "dist-cjs/index.js",
        "module": "dist-es/index.js",
        "name": "sample-package",
        "typings": "dist-types/src/index.d.ts",
        "version": "1.0.0",
      }
    `);

    expect(
      await fs.readJson(
        path.join(modularRoot, 'dist', 'sample-view', 'package.json'),
      ),
    ).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {
          "react": "^17.0.2",
        },
        "files": Array [
          "/dist-cjs",
          "/dist-es",
          "/dist-types",
          "README.md",
        ],
        "license": "UNLICENSED",
        "main": "dist-cjs/sample-view.cjs.js",
        "modular": Object {
          "type": "view",
        },
        "module": "dist-es/sample-view.es.js",
        "name": "sample-view",
        "typings": "dist-types/src/index.d.ts",
        "version": "1.0.0",
      }
    `);

    expect(
      await fs.readJson(
        path.join(
          modularRoot,
          'dist',
          'nested/sample-nested-package',
          'package.json',
        ),
      ),
    ).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {},
        "files": Array [
          "/dist-cjs",
          "/dist-es",
          "/dist-types",
          "README.md",
        ],
        "license": "UNLICENSED",
        "main": "dist-cjs/nested-sample-package.cjs.js",
        "module": "dist-es/nested-sample-package.es.js",
        "name": "@nested/sample-package",
        "typings": "dist-types/src/index.d.ts",
        "version": "1.0.0",
      }
    `);

    expect(tree(path.join(modularRoot, 'dist'))).toMatchInlineSnapshot(`
      "dist
      ├─ nested
      │  └─ sample-nested-package
      │     ├─ README.md #1jv3l2q
      │     ├─ dist-cjs
      │     │  ├─ nested-sample-package.cjs.js #kv2xzp
      │     │  └─ nested-sample-package.cjs.js.map #j26x67
      │     ├─ dist-es
      │     │  ├─ nested-sample-package.es.js #40jnpo
      │     │  └─ nested-sample-package.es.js.map #11g8lh9
      │     ├─ dist-types
      │     │  └─ src
      │     │     └─ index.d.ts #f68aj
      │     └─ package.json
      ├─ sample-package
      │  ├─ README.md #1jv3l2q
      │  ├─ dist-cjs
      │  │  ├─ index.js #rq9uxe
      │  │  └─ index.js.map #ys8x0i
      │  ├─ dist-es
      │  │  ├─ index.js #1gjntzw
      │  │  └─ index.js.map #b17359
      │  ├─ dist-types
      │  │  └─ src
      │  │     └─ index.d.ts #f68aj
      │  └─ package.json
      └─ sample-view
         ├─ README.md #11adaka
         ├─ dist-cjs
         │  ├─ sample-view.cjs.js #8jw6cg
         │  └─ sample-view.cjs.js.map #130r3z8
         ├─ dist-es
         │  ├─ sample-view.es.js #1ctbbz8
         │  └─ sample-view.es.js.map #12deywy
         ├─ dist-types
         │  └─ src
         │     └─ index.d.ts #1vloh7q
         └─ package.json"
    `);
  });
});
