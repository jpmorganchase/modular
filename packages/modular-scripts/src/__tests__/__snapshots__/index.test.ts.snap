// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modular-scripts WHEN building a view THEN outputs the correct output cjs file 1`] = `
"'use strict';

var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return n;
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

function SampleView() {
  return /*#__PURE__*/React__namespace.createElement(\\"div\\", {
    \\"data-testid\\": \\"test-this\\"
  }, \\"this is a modular view\\");
}

module.exports = SampleView;
//# sourceMappingURL=sample-view.cjs.js.map
"
`;

exports[`modular-scripts WHEN building a view THEN outputs the correct output cjs map file 1`] = `
Object {
  "file": "sample-view.cjs.js",
  "mappings": ";;;;;;;;;;;;;;;;;;;;;;;;;;AAEe,SAASA,UAAT,GAAmC;AAChD,sBAAOC,iEAAP;AACD;;;;",
  "names": Array [
    "SampleView",
    "React",
  ],
  "sources": Array [
    "../src/index.tsx",
  ],
  "sourcesContent": Array [
    "import * as React from 'react';

export default function SampleView(): JSX.Element {
  return <div>This is SampleView</div>;
}
",
  ],
  "version": 3,
}
`;
